Python: Function/Class Naming Convention (Syntax)

dilanbhalla opened the pull request:
A minor addition for maintaining standard coding practices when naming functions/classes in python. The two queries simply warn against function names that are capitalized or classes that are not. It is a small/simple addition but important practice that is especially helpful for readability purposes when sharing code with peers.

RasmusWL commented on python/ql/src/Functions/NamingConventionsFunctions.qhelp:
```diff
@@ -0,0 +1,30 @@
+
+<overview>
+<p>A function name that begins with an uppercase letter does not follow standard
+naming conventions. This decreases code readability. For example, <code>Jump</code>.
```
```suggestion
naming conventions. This decreases code readability. For example, <code>def Jump():</code>.
```

RasmusWL commented on python/ql/src/Functions/NamingConventionsFunctions.qhelp:
```diff
@@ -0,0 +1,30 @@
+<recommendation>
+
+<p>
+Write the function name beginning with an lowercase letter. For example, <code>jump</code>.
```
```suggestion
Write the function name beginning with an lowercase letter. For example, <code>def jump():</code>.
```

RasmusWL commented on python/ql/src/Classes/NamingConventionsClasses.ql:
```diff
@@ -0,0 +1,17 @@
+ * @problem.severity recommendation
+ * @precision medium
+ * @id python/misnamed-type
+ * @tags maintainability
```
```suggestion
 * @tags readability
```

RasmusWL commented on python/ql/src/Functions/NamingConventionsFunctions.ql:
```diff
@@ -0,0 +1,17 @@
+ * @problem.severity recommendation
+ * @precision medium
+ * @id python/misnamed-function
+ * @tags maintainability
```
```suggestion
 * @tags readability
```

dilanbhalla commented:
Hello,

Thank you for the fixes. Looked great to me so I committed them, let me know if there is anything else you would like me to do! (I added these before I was informed that I should submit new queries under “experimental,” so if you would like me to move these there I can definitely do that as well)

Best,
Dilan

> On Jun 26, 2020, at 4:42 AM, Rasmus Wriedt Larsen <notifications@github.com> wrote:
> 
> ﻿
> @RasmusWL requested changes on this pull request.
> 
> Thanks for your submission @dilanbhalla 👍
> 
> Overall I think these cases are better handled by a linter (and pylint does mark these with the invalid-name rule), but I also don't see anything wrong with adding these queries. There is a few minor changes I would like to see just quickly looking over the code.
> 
> @tausbn how do you feel about adding these queires?
> 
> In python/ql/src/Functions/NamingConventionsFunctions.qhelp:
> 
> > @@ -0,0 +1,30 @@
> +<!DOCTYPE qhelp PUBLIC
> +  "-//Semmle//qhelp//EN"
> +  "qhelp.dtd">
> +<qhelp>
> +
> +
> +<overview>
> +<p>A function name that begins with an uppercase letter does not follow standard
> +naming conventions. This decreases code readability. For example, <code>Jump</code>.
> ⬇️ Suggested change
> -naming conventions. This decreases code readability. For example, <code>Jump</code>.
> +naming conventions. This decreases code readability. For example, <code>def Jump():</code>.
> In python/ql/src/Functions/NamingConventionsFunctions.qhelp:
> 
> > +<!DOCTYPE qhelp PUBLIC
> +  "-//Semmle//qhelp//EN"
> +  "qhelp.dtd">
> +<qhelp>
> +
> +
> +<overview>
> +<p>A function name that begins with an uppercase letter does not follow standard
> +naming conventions. This decreases code readability. For example, <code>Jump</code>.
> +</p>
> +
> +</overview>
> +<recommendation>
> +
> +<p>
> +Write the function name beginning with an lowercase letter. For example, <code>jump</code>.
> ⬇️ Suggested change
> -Write the function name beginning with an lowercase letter. For example, <code>jump</code>.
> +Write the function name beginning with an lowercase letter. For example, <code>def jump():</code>.
> In python/ql/src/Classes/NamingConventionsClasses.ql:
> 
> > @@ -0,0 +1,17 @@
> +/**
> + * @name Misnamed class
> + * @description A class name that begins with a lowercase letter decreases readability.
> + * @kind problem
> + * @problem.severity recommendation
> + * @precision medium
> + * @id python/misnamed-type
> + * @tags maintainability
> ⬇️ Suggested change
> - * @tags maintainability
> + * @tags readability
> In python/ql/src/Functions/NamingConventionsFunctions.ql:
> 
> > @@ -0,0 +1,17 @@
> +/**
> + * @name Misnamed function
> + * @description A function name that begins with an uppercase letter decreases readability.
> + * @kind problem
> + * @problem.severity recommendation
> + * @precision medium
> + * @id python/misnamed-function
> + * @tags maintainability
> ⬇️ Suggested change
> - * @tags maintainability
> + * @tags readability
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

dilanbhalla commented:
Hello,

I just moved this query to experimental in case it would like to be reviewed there first as well. Thank you again and let me know if I need to include anything else!

Best,
Dilan

> On Jul 1, 2020, at 9:34 AM, Dilan Bhalla <dbhalla21@berkeley.edu> wrote:
> 
> ﻿Hello,
> 
> Thank you for the fixes. Looked great to me so I committed them, let me know if there is anything else you would like me to do! (I added these before I was informed that I should submit new queries under “experimental,” so if you would like me to move these there I can definitely do that as well)
> 
> Best,
> Dilan
> 
>>> On Jun 26, 2020, at 4:42 AM, Rasmus Wriedt Larsen <notifications@github.com> wrote:
>>> 
>> ﻿
>> @RasmusWL requested changes on this pull request.
>> 
>> Thanks for your submission @dilanbhalla 👍
>> 
>> Overall I think these cases are better handled by a linter (and pylint does mark these with the invalid-name rule), but I also don't see anything wrong with adding these queries. There is a few minor changes I would like to see just quickly looking over the code.
>> 
>> @tausbn how do you feel about adding these queires?
>> 
>> In python/ql/src/Functions/NamingConventionsFunctions.qhelp:
>> 
>> > @@ -0,0 +1,30 @@
>> +<!DOCTYPE qhelp PUBLIC
>> +  "-//Semmle//qhelp//EN"
>> +  "qhelp.dtd">
>> +<qhelp>
>> +
>> +
>> +<overview>
>> +<p>A function name that begins with an uppercase letter does not follow standard
>> +naming conventions. This decreases code readability. For example, <code>Jump</code>.
>> ⬇️ Suggested change
>> -naming conventions. This decreases code readability. For example, <code>Jump</code>.
>> +naming conventions. This decreases code readability. For example, <code>def Jump():</code>.
>> In python/ql/src/Functions/NamingConventionsFunctions.qhelp:
>> 
>> > +<!DOCTYPE qhelp PUBLIC
>> +  "-//Semmle//qhelp//EN"
>> +  "qhelp.dtd">
>> +<qhelp>
>> +
>> +
>> +<overview>
>> +<p>A function name that begins with an uppercase letter does not follow standard
>> +naming conventions. This decreases code readability. For example, <code>Jump</code>.
>> +</p>
>> +
>> +</overview>
>> +<recommendation>
>> +
>> +<p>
>> +Write the function name beginning with an lowercase letter. For example, <code>jump</code>.
>> ⬇️ Suggested change
>> -Write the function name beginning with an lowercase letter. For example, <code>jump</code>.
>> +Write the function name beginning with an lowercase letter. For example, <code>def jump():</code>.
>> In python/ql/src/Classes/NamingConventionsClasses.ql:
>> 
>> > @@ -0,0 +1,17 @@
>> +/**
>> + * @name Misnamed class
>> + * @description A class name that begins with a lowercase letter decreases readability.
>> + * @kind problem
>> + * @problem.severity recommendation
>> + * @precision medium
>> + * @id python/misnamed-type
>> + * @tags maintainability
>> ⬇️ Suggested change
>> - * @tags maintainability
>> + * @tags readability
>> In python/ql/src/Functions/NamingConventionsFunctions.ql:
>> 
>> > @@ -0,0 +1,17 @@
>> +/**
>> + * @name Misnamed function
>> + * @description A function name that begins with an uppercase letter decreases readability.
>> + * @kind problem
>> + * @problem.severity recommendation
>> + * @precision medium
>> + * @id python/misnamed-function
>> + * @tags maintainability
>> ⬇️ Suggested change
>> - * @tags maintainability
>> + * @tags readability
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub, or unsubscribe.

tausbn commented on python/ql/src/experimental/Functions/NamingConventionsFunctions.ql:
```diff
@@ -0,0 +1,17 @@
+
+import python
+
+from Function f
+where
+  f.inSource() and
+  not f.getName().substring(0, 1).toLowerCase() = f.getName().substring(0, 1)
+select f, "Function names should start in lowercase."
```
```suggestion
from Function f, string first_char
where
  f.inSource() and
  first_char = f.getName().prefix(1) and
  not first_char = first_char.toLowerCase()
select f, "Function names should start in lowercase."
```
I would suggest rewriting the query as follows, to eliminate the repeated `substring`, and to (hopefully) make the intent clearer. A similar comment applies to the other query.

dilanbhalla replied:
This has been changed to the method you have suggested here. With regards to the first suggestion you made, I modified the queries so that it only enters the select statement if the class/function is the only one in its file that has the incorrect case. On lgtm this reduced the number of results from ~15 per project to ~7 per project, eliminating many false positives I believe. Let me know if you think that makes sense or if a different approach would be better. Thanks again!

dilanbhalla commented on python/ql/src/experimental/Classes/NamingConventionsClasses.qhelp:
```diff
@@ -0,0 +1,30 @@
+
+<li>
+  Guido van Rossum, Barry Warsaw, Nick Coghlan <em>PEP 8 -- Style Guide for Python Code<em>
+  <a href="https://www.python.org/dev/peps/pep-0008/#class-names"</a>
```
Fixed, as well as the other comments you mentioned below. Thank you for all of the suggestions!

tausbn commented on python/ql/src/experimental/Classes/NamingConventionsClasses.qhelp:
```diff
@@ -0,0 +1,30 @@
+<references>
+
+<li>
+  Guido van Rossum, Barry Warsaw, Nick Coghlan <em>PEP 8 -- Style Guide for Python Code<em>
```
```suggestion
  Guido van Rossum, Barry Warsaw, Nick Coghlan <em>PEP 8 -- Style Guide for Python Code</em>
```

tausbn commented on python/ql/src/experimental/Functions/NamingConventionsFunctions.qhelp:
```diff
@@ -0,0 +1,30 @@
+<references>
+
+<li>
+  Guido van Rossum, Barry Warsaw, Nick Coghlan <em>PEP 8 -- Style Guide for Python Code<em>
```
```suggestion
  Guido van Rossum, Barry Warsaw, Nick Coghlan <em>PEP 8 -- Style Guide for Python Code</em>
```

tausbn commented on python/ql/src/experimental/Classes/NamingConventionsClasses.ql:
```diff
@@ -0,0 +1,23 @@
+from Class c, string first_char
+where
+  c.inSource() and
+  first_char = c.getName().prefix(1) and
+  not first_char = first_char.toUpperCase() and
```
I think it would help readability (and avoid repetition) if this was factored out into a separate predicate, for instance
```ql
predicate lower_case_class(File file, Class c) {
  file = c.getLocation().getFile()
  exists(string first_char | 
    first_char = c.getName().prefix(1) and
    not first_char = first_char.toUpperCase()
  )
}
```
(untested)

That way, you can reference this twice in the query.

tausbn commented on python/ql/src/experimental/Functions/NamingConventionsFunctions.ql:
```diff
@@ -0,0 +1,23 @@
+from Function f, string first_char
+where
+  f.inSource() and
+  first_char = f.getName().prefix(1) and
+  not first_char = first_char.toLowerCase() and
```
Same comment as above, except with `Function` instead of `Class` in the helper predicate.

tausbn reviewed (changes requested):
Hello! Thank you for your patience. I have made a few suggestions, two for the `.qhelp` files which had incorrect closing tags, and two for the queries themselves. I think this looks pretty close to being mergeable.

(And as you can tell, we're slowly getting back to reviewing external contributions again. We'll be revisiting the rest of your contributions in the near future.)

dilanbhalla commented:
Thanks for all of your comments! Incorporated all of the fixes.

tausbn reviewed (changes requested):
I think this looks good to merge. However, it's currently failing the autoformatting CI check, and that's preventing us from merging. You'll need to autoformat the two `.ql` files and push the changes.

dilanbhalla commented:
Fixed

tausbn commented:
The CI is still complaining:
```
ql/python/ql/src/experimental/Functions/NamingConventionsFunctions.ql would change by autoformatting.
```

dilanbhalla commented:
That's super weird. I autoformatted and pushed changes for both but looks like only one changed looking back at the commit. I autoformatted the second and pushed, hopefully it is okay now.

tausbn commented:
> @tausbn how do you feel about adding these queires?

I think these are fine to merge in their present form. With the addition of looking at the surrounding context and the functions and classes defined there, I think this might actually be a useful check to have. I don't know of any linters that do this, at any rate.

RasmusWL reviewed (approved):
Thanks, this looks fine to me :+1:
